### DISCUSSION TITLE
Formalize policy for releases

### DISCUSSION THREAD

**Discussion by philippjfr** [bokeh/bokeh#11702] [Build / CI]
In the context of the Bokeh 2.4 release we started discussing what the policy around release candidates, feature freezes, etc. should be. Specifically in the past we have had a policy of declaring a feature freeze one week ahead of release which would be accompanied by an RC. Subsequently we would only merge bug/regression fixes followed by further release candidates. This policy has never really been formalized and so it was only partially adhered to. Therefore I wanted to start a discussion to work out a formal policy.

There's three distinct groups of stakeholders with different needs affected here:

1. The developers/contributors of Bokeh would generally like more flexibility in their releases but are of course invested in regression free releases.
2. The users of Bokeh want stable and bug free releases and time to test their code against an RC to catch issues before a release.
3. Developers of Bokeh extensions and tools that build on Bokeh need time to evaluate an RC against their downstream libraries and report issues before they make it into a release that they have to support.

Whatever policy we come up with should give both users and downstream developers sufficient lead time to test an RC and report any issues. This should also be accompanied by a formal process for publicly announcing RC releases and notifying relevant stakeholders.

Some questions we should answer:

1. How long between feature freeze and the availability of an RC and the final release (in the absence of any release blockers being discovered)?
2. How should RC releases be publicized and should we solicit an explicit list of stakeholders which should be notified?
3. Should there be a process for getting a feature in which arrives during a feature freeze? Does this reset the feature freeze period?
4. How strict should these processes be and who gets to decide to make exceptions?

**Comment by jbednar** [comment]
1. My vote for 1 would be one week for a minor release like 2.4 and two to three business days for a patch release like 2.3.3. Often a patch release is addressing bugs in the previous release, so it's important to be able to get that kind out quickly. Minor releases are more feature laden and thus presumably less time critical. 
2. I think it's worth notifying package authors for things that explicitly depend on bokeh, e.g. Panel, HoloViews, Pandas-Bokeh, Chartify, Dask. 
3. I'm less worried about new features coming in than I'm worried about API changes. So I would focus on regulating any potential API breakages rather than features, personally!
4. I assume that whoever is managing the release would be in charge! That person should of course take input, but realistically, if they have taken on the responsibility of making the release, I think they have to have the freedom to make exceptions when they believe it is appropriate.

**Comment by bryevdv** [comment]
> My vote for 1 would be one week for a minor release like 2.4 and two to three business days for a patch release like 2.3.3. Often a patch release is addressing bugs in the previous release, so it's important to be able to get that kind out quickly. Minor releases are more feature laden and thus presumably less time critical.

I'd like to propose that major/minor releases happen on Wednesdays, right after the meeting. That way the attendees can be polled for 👍 / 👎 to go ahead. Given that, I would propose initial RC1s happen on the previous Thursday. 

Here is what I would point out though: in order for there to be confidence that "RC1" does in fact have a good chance of not needing more work, testing for downstream projects needs to happen *before that point*. i.e. the goal needs to be that an "RC1" is "ready to release" not that is is "ready to test". 

For point releases I would be ok with (weekday) 48hrs from RC to release. Now that we can actually issue point releases on old branches, I tool a look back an we have been reasonably conservative with both size and content thus far. 

> I think it's worth notifying package authors for things that explicitly depend on bokeh, e.g. Panel, HoloViews, Pandas-Bokeh, Chartify, Dask.

Sure but notify how? My only real concern/requirement is that it be something we can move towards automating. (emails, discourse posts, slack posts, issue creation on specified trackers, whatever) 

> I assume that whoever is managing the release in charge! 

Yes, and I think we should consider making this a rotating responsibility, if only to reduce the bus-factor

**Reply by bryevdv** [reply]
> How long between feature freeze and the availability of an RC and the final release 

Let's define "feature freeze". I'd like it to mean nothing new is added to the milestone, i.e a period of time when only burn down is allowed. Keeping things simple, for major/minor releases we could do another week before. Off the top of my head, then a "normal" timeline might look like:

* first wednesday meeting: 👍 / 👎  for "we are now in freeze". If "yes" then only burndown is allowed from that point:
* second wednesday meeting: 👍 / 👎 to end burndown period and issue RC1. If "yes" then anything left in the milestone is moved out:
* third wednesday meeting: 👍 / 👎 to release 

If any of the decisions above are "no" or if new critical work is discovered that must be added to the milestone, then that extends the current period to the following wednesday to reconsider the next stage.

**Comment by jbednar** [comment]
> in order for there to be confidence that "RC1" does in fact have a good chance of not needing more work, testing for downstream projects needs to happen before that point. i.e. the goal needs to be that an "RC1" is "ready to release" not that is is "ready to test".

Hmm. To me, that's what an "RC" already is: ready to test manually. I.e., all automated tests are passing, all features are in, basic manual tests pass, could be ready to go; now let's all test and find out if it really is!  As an external project maintainer, I need to know when such a point has been reached in Bokeh so that it's appropriate to run through a manual checklist or even that it's worth investigating failures on CI that we could otherwise think are just due to things in progress. An alpha release isn't one that I'd think is worth that effort; sometimes in my projects we have a dozen or more alpha releases for a given public release.  But if an RC appears, we should get into gear and start seeing what work we have.

In any case, if you don't want to call that stage RC, would you call it "gamma" or some other level that indicates not quite RC but ready for external testers? If so, that's fine; I'm not wedded to those particular characters...
